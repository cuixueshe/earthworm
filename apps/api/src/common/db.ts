import { Logger } from "@nestjs/common";
import { DefaultLogger, LogWriter } from "drizzle-orm";
import { drizzle } from "drizzle-orm/postgres-js";
import * as postgres from "postgres";

import { schemas } from "@earthworm/schema";

let connection: postgres.Sql;

async function createConnection() {
  return postgres(process.env.DATABASE_URL ?? "");
}

export async function endDB() {
  if (connection) {
    await connection.end();
    connection = null;
  }
}

export async function setupDB() {
  if (connection) return;

  const logger = new Logger("DB");

  class CustomDbLogWriter implements LogWriter {
    write(message: string) {
      logger.verbose(message);
    }
  }

  logger.debug(`Connecting to ${process.env.DATABASE_URL}`);
  logger.debug(`SECRET: ${process.env.SECRET}`);

  connection = await createConnection();

  return drizzle(connection, {
    schema: schemas,
    logger: new DefaultLogger({ writer: new CustomDbLogWriter() }),
  });
}
