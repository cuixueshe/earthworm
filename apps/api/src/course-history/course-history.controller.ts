import { Controller, Get, Param, UseGuards } from "@nestjs/common";
import { ApiBearerAuth, ApiOperation, ApiTags } from "@nestjs/swagger";

import { AuthGuard } from "../guards/auth.guard";
import { User, UserEntity } from "../user/user.decorators";
import { CourseHistoryService } from "./course-history.service";

@ApiBearerAuth()
@ApiTags("CourseHistory")
@Controller("course-history")
export class CourseHistoryController {
  constructor(private readonly courseHistoryService: CourseHistoryService) {}

  @ApiOperation({
    summary: "获取登陆用户的所有课程历史记录",
  })
  @UseGuards(AuthGuard)
  @Get("")
  courseCompletionCount(@User() user: UserEntity) {
    return this.courseHistoryService.findAll(user.userId);
  }

  @UseGuards(AuthGuard)
  @Get(":coursePackId")
  getCoursePackHistory(@User() user: UserEntity, @Param("coursePackId") coursePackId: string) {
    return this.courseHistoryService.findByCoursePackId(user.userId, coursePackId);
  }
}
