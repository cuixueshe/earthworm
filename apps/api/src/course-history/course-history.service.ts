import { courseHistory } from '@earthworm/schema';
import { Inject, Injectable } from '@nestjs/common';
import { and, eq } from 'drizzle-orm';
import { UserEntity } from 'src/user/user.decorators';
import { DB, DbType } from '../global/providers/db.provider';

@Injectable()
export class CourseHistoryService {
  constructor(@Inject(DB) private db: DbType) {}

  async findOne(userId: number, courseId: number) {
    return await this.db
      .select()
      .from(courseHistory)
      .where(
        and(
          eq(courseHistory.userId, userId),
          eq(courseHistory.courseId, courseId),
        ),
      );
  }

  async create(userId: number, courseId: number) {
    await this.db.insert(courseHistory).values({
      courseId,
      userId,
      completionCount: 1,
    });
  }

  async updateCompletionCount(userId: number, courseId: number, count: number) {
    await this.db
      .update(courseHistory)
      .set({
        completionCount: count + 1,
      })
      .where(
        and(
          eq(courseHistory.userId, userId),
          eq(courseHistory.courseId, courseId),
        ),
      );
  }

  async setCompletionCount(userId: number, courseId: number) {
    const result = await this.findOne(userId, courseId);
    if (result && result.length) {
      this.updateCompletionCount(userId, courseId, result[0].completionCount);
    } else {
      this.create(userId, courseId);
    }
  }

  async findAll(user: UserEntity) {
    return await this.db
      .select()
      .from(courseHistory)
      .where(eq(courseHistory.userId, user.userId));
  }
}
