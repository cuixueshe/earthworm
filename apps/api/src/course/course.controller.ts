import { Controller, Get, Param, Post, UseGuards } from "@nestjs/common";
import { ApiBearerAuth, ApiOperation, ApiTags } from "@nestjs/swagger";

import { AuthGuard } from "../auth/auth.guard";
import { User, UserEntity } from "../user/user.decorators";
import { CourseService } from "./course.service";

@ApiTags("Courses")
@Controller("courses")
export class CourseController {
  constructor(private courseService: CourseService) {}

  @ApiOperation({
    summary: "返回第一课课程信息，用于游客体验",
  })
  @Get("try")
  tryCourse() {
    // 专门让游客体验的接口 不需要做 auth 验证
    return this.courseService.tryCourse();
  }

  @ApiOperation({
    summary: "根据课程ID返回课程内容",
  })
  @ApiBearerAuth()
  @UseGuards(AuthGuard)
  @Get(":courseId")
  findOne(@Param("courseId") courseId: number) {
    return this.courseService.find(courseId);
  }

  @ApiOperation({
    summary: "获取所有课程的信息",
  })
  @ApiBearerAuth()
  @UseGuards(AuthGuard)
  @Get("")
  findAll() {
    return this.courseService.findAll();
  }

  @ApiOperation({
    summary: "查找指定课程ID（courseId）之后的下一个课程，并返回该课程的信息",
  })
  @ApiBearerAuth()
  @UseGuards(AuthGuard)
  @Get(":courseId/next")
  findNext(@Param("courseId") courseId: number) {
    return this.courseService.findNext(courseId);
  }

  @ApiOperation({
    summary: "完成指定课程，更新课程进度，完成次数，排行榜，返回下一课内容",
  })
  @ApiBearerAuth()
  @UseGuards(AuthGuard)
  @Post(":courseId/complete")
  async completeCourse(@User() user: UserEntity, @Param("courseId") courseId: number) {
    const result = await this.courseService.completeCourse(user, courseId);
    return result;
  }
}
