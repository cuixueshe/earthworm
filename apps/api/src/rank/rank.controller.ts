import { Controller, Get, Param, UseGuards } from "@nestjs/common";
import { ApiBearerAuth, ApiOperation, ApiTags } from "@nestjs/swagger";

import type { RankPeriodAlias } from "./rank.service";
import { AuthGuard, UncheckAuth } from "../guards/auth.guard";
import { User, UserEntity } from "../user/user.decorators";
import { RankPeriod, RankService } from "./rank.service";

@ApiBearerAuth()
@ApiTags("Rank")
@Controller("rank")
export class RankController {
  constructor(private readonly rankService: RankService) {}

  @ApiOperation({
    summary: "获取排行榜信息",
  })
  @UncheckAuth()
  @UseGuards(AuthGuard)
  @Get("progress/:period")
  getRankList(
    @User() user: UserEntity,
    @Param("period") period: RankPeriodAlias = RankPeriod.WEEKLY,
  ) {
    return this.rankService.getRankList(user, period);
  }
}
