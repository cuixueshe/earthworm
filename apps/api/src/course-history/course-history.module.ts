import { Module, forwardRef } from '@nestjs/common';
import { CourseHistoryService } from './course-history.service';
import { CourseHistoryController } from './course-history.controller';
import { CourseModule } from '../course/course.module';

@Module({
  imports: [forwardRef(() => CourseModule)],
  controllers: [CourseHistoryController],
  providers: [CourseHistoryService],
  exports: [CourseHistoryService],
})
export class CourseHistoryModule {}
