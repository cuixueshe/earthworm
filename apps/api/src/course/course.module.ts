import { Module } from '@nestjs/common';
import { CourseService } from './course.service';
import { CourseController } from './course.controller';
import { GlobalModule } from '../global/global.module';
import { UserProgressService } from '../user-progress/user-progress.service';
import { RankService } from '../rank/rank.service';
import { CourseHistoryService } from '../course-history/course-history.service';
import { UserLearnRecordService } from '../user-learn-record/user-learn-record.service';

@Module({
  imports: [GlobalModule],
  providers: [
    CourseService,
    UserProgressService,
    RankService,
    CourseHistoryService,
    UserLearnRecordService,
  ],
  controllers: [CourseController],
  exports: [CourseService],
})
export class CourseModule {}
