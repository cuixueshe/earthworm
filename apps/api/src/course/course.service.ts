import { Inject, Injectable, NotFoundException } from "@nestjs/common";
import { and, asc, eq, gt } from "drizzle-orm";

import { course, statement } from "@earthworm/schema";
import { CourseHistoryService } from "../course-history/course-history.service";
import { DB, DbType } from "../global/providers/db.provider";
import { RankService } from "../rank/rank.service";
import { UserCourseProgressService } from "../user-course-progress/user-course-progress.service";
import { UserLearnRecordService } from "../user-learn-record/user-learn-record.service";

@Injectable()
export class CourseService {
  constructor(
    @Inject(DB) private db: DbType,
    private readonly rankService: RankService,
    private readonly courseHistoryService: CourseHistoryService,
    private readonly userLearnRecordService: UserLearnRecordService,
    private readonly userCourseProgressService: UserCourseProgressService,
  ) {}

  async find(coursePackId: number, courseId: number) {
    const courseEntity = await this.db.query.course.findFirst({
      where: and(eq(course.id, courseId), eq(course.coursePackId, coursePackId)),
      with: {
        statements: {
          columns: {
            id: false,
          },
          orderBy: [asc(statement.order)],
        },
      },
    });

    if (!courseEntity) {
      throw new NotFoundException(
        `CoursePack with ID ${coursePackId} and CourseId with ID ${courseId} not found`,
      );
    }

    return courseEntity;
  }

  async findWithUserProgress(coursePackId: number, courseId: number, userId: string) {
    const courseEntity = await this.find(coursePackId, courseId);

    const statementIndex = await this.userCourseProgressService.findStatement(
      userId,
      coursePackId,
      courseId,
    );

    return { ...courseEntity, statementIndex };
  }

  async findNext(coursePackId: number, courseId: number) {
    const result = await this._findNext(coursePackId, courseId);

    if (!result) {
      throw new NotFoundException(
        `Can't find the next course -> coursePackId: ${coursePackId} courseId: ${courseId}`,
      );
    }

    return result;
  }

  private async _findNext(coursePackId: number, courseId: number) {
    const nextCourse = await this.db.query.course.findFirst({
      where: and(gt(course.id, courseId), eq(course.coursePackId, coursePackId)),
    });

    return nextCourse;
  }

  async completeCourse(userId: string, coursePackId: number, courseId: number) {
    const nextCourse = await this._findNext(coursePackId, courseId);

    if (userId) {
      await this.rankService.userFinishCourse(userId);
      await this.courseHistoryService.upsert(userId, coursePackId, courseId);
      await this.userLearnRecordService.upsert(userId);
      nextCourse &&
        (await this.userCourseProgressService.upsert(userId, coursePackId, nextCourse.id, 0));
    }

    return {
      nextCourse: await this._findNext(coursePackId, courseId),
    };
  }
}
