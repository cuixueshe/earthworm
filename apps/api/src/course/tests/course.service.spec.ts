import { NotFoundException } from "@nestjs/common";
import { Test } from "@nestjs/testing";

import type { DbType } from "../../global/providers/db.provider";
import { insertCourse, insertCoursePack, insertStatement } from "../../../test/fixture/db";
import { cleanDB, testImportModules } from "../../../test/helper/utils";
import { endDB } from "../../common/db";
import { CourseHistoryService } from "../../course-history/course-history.service";
import { DB } from "../../global/providers/db.provider";
import { RankService } from "../../rank/rank.service";
import { UserCourseProgressService } from "../../user-course-progress/user-course-progress.service";
import { UserLearnRecordService } from "../../user-learn-record/user-learn-record.service";
import { CourseService } from "../course.service";

describe("course service", () => {
  let db: DbType;
  let courseService: CourseService;
  let rankService: RankService;
  let courseHistoryService: CourseHistoryService;
  let userLearnRecordService: UserLearnRecordService;
  let userCourseProgressService: UserCourseProgressService;

  beforeAll(async () => {
    const testHelper = await setupTesting();
    await setupDatabaseData(testHelper.db);

    db = testHelper.db;
    courseService = testHelper.courseService;
    rankService = testHelper.rankService;
    courseHistoryService = testHelper.courseHistoryService;
    userLearnRecordService = testHelper.userLearnRecordService;
    userCourseProgressService = testHelper.UserCourseProgressService;
  });
  beforeEach(async () => {
    await cleanDB(db);
  });

  afterAll(async () => {
    await cleanDB(db);
    await endDB();
  });

  afterEach(async () => {
    jest.clearAllMocks();
  });

  describe("find", () => {
    it("should return a course with the given coursePackId and courseId", async () => {
      const { coursePackId, courseEntityFirst } = await setupDBData(db);

      const result = await courseService.find(coursePackId, courseEntityFirst.id);

      expect(result).toMatchSnapshot();
    });

    it("should throw NotFoundException if the course does not exist", async () => {
      await expect(courseService.find(2, 1)).rejects.toThrow(NotFoundException);
    });
  });

  describe("findWithUserProgress", () => {
    it("should return a course with user progress information", async () => {
      const { coursePackId, courseEntityFirst, userId } = await setupDBData(db);

      const result = await courseService.findWithUserProgress(
        coursePackId,
        courseEntityFirst.id,
        userId,
      );

      expect(result).toHaveProperty("statementIndex");
    });
  });

  describe("findNext", () => {
    it("should return the next course", async () => {
      const { coursePackId, courseEntityFirst, courseEntitySecond } = await setupDBData(db);

      const result = await courseService.findNext(coursePackId, courseEntityFirst.id);

      expect(result).toEqual(courseEntitySecond);
    });

    it("should throw NotFoundException if there is no next course", async () => {
      const { coursePackId, courseEntitySecond } = await setupDBData(db);

      await expect(courseService.findNext(coursePackId, courseEntitySecond.id)).rejects.toThrow(
        NotFoundException,
      );
    });
  });

  describe("completeCourse", () => {
    it("should perform actions to complete a course for a user with userId and return the next course", async () => {
      const { userId, courseEntityFirst, coursePackId } = await setupDBData(db);

      const result = await courseService.completeCourse(userId, coursePackId, courseEntityFirst.id);

      expect(result).toHaveProperty("nextCourse");
      expect(rankService.userFinishCourse).toHaveBeenCalled();
      expect(courseHistoryService.upsert).toHaveBeenCalled();
      expect(userLearnRecordService.upsert).toHaveBeenCalled();
      expect(userCourseProgressService.upsert).toHaveBeenCalled();
    });

    it("should perform actions to complete a course and return the next course when have not userId", async () => {
      const { courseEntityFirst, coursePackId } = await setupDBData(db);

      const result = await courseService.completeCourse("", coursePackId, courseEntityFirst.id);

      expect(result).toHaveProperty("nextCourse");
      expect(rankService.userFinishCourse).not.toHaveBeenCalled();
      expect(courseHistoryService.upsert).not.toHaveBeenCalled();
      expect(userLearnRecordService.upsert).not.toHaveBeenCalled();
      expect(userCourseProgressService.upsert).not.toHaveBeenCalled();
    });

    it("should not have nextCourse when not exist next course", async () => {
      const { courseEntitySecond, coursePackId } = await setupDBData(db);

      const result = await courseService.completeCourse("", coursePackId, courseEntitySecond.id);

      expect(result.nextCourse).toBeUndefined();
    });
  });
});

async function setupDatabaseData(db: DbType) {
  await cleanDB(db);
}

async function setupTesting() {
  const mockRankService = {
    userFinishCourse: jest.fn(),
  };
  const mockCourseHistoryService = {
    upsert: jest.fn(),
  };
  const mockUserLearnRecordService = {
    upsert: jest.fn(),
  };

  const mockUserCourseProgressService = {
    upsert: jest.fn(),
    findStatement: () => 1,
  };

  const moduleRef = await Test.createTestingModule({
    imports: testImportModules,
    providers: [
      CourseService,
      { provide: RankService, useValue: mockRankService },
      { provide: CourseHistoryService, useValue: mockCourseHistoryService },
      { provide: UserLearnRecordService, useValue: mockUserLearnRecordService },
      { provide: UserCourseProgressService, useValue: mockUserCourseProgressService },
    ],
  }).compile();

  return {
    courseService: moduleRef.get<CourseService>(CourseService),
    UserCourseProgressService: moduleRef.get<UserCourseProgressService>(UserCourseProgressService),
    rankService: moduleRef.get<RankService>(RankService),
    courseHistoryService: moduleRef.get<CourseHistoryService>(CourseHistoryService),
    userLearnRecordService: moduleRef.get<UserLearnRecordService>(UserLearnRecordService),
    db: moduleRef.get<DbType>(DB),
    moduleRef,
  };
}

async function setupDBData(db: DbType) {
  const userId = "cxr";
  const coursePackEntity = await insertCoursePack(db);
  const courseEntityFirst = await insertCourse(db, coursePackEntity.id, "第一课");
  const courseEntitySecond = await insertCourse(db, coursePackEntity.id, "第二课");
  const statementEntityFirst = await insertStatement(db, courseEntityFirst.id, 1);
  const statementEntitySecond = await insertStatement(db, courseEntityFirst.id, 2);

  return {
    userId,
    coursePackId: coursePackEntity.id,
    courseEntityFirst,
    courseEntitySecond,
    statementEntityFirst,
    statementEntitySecond,
  };
}
