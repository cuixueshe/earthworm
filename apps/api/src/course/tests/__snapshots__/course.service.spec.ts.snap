// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`course service find should return a course with the given coursePackId and courseId 1`] = `
{
  "coursePackId": 1,
  "id": 1,
  "statements": [
    {
      "chinese": "你好",
      "courseId": 1,
      "english": "hello",
      "order": 1,
      "soundmark": "nihao",
    },
    {
      "chinese": "你好",
      "courseId": 1,
      "english": "hello",
      "order": 2,
      "soundmark": "nihao",
    },
  ],
  "title": "第一课",
}
`;
